synapse/rest/client/v2_alpha/account.py

```diff
@@ -485,12 +542,126 @@ def on_POST(self, request):
         if existing_user_id is not None:
             raise SynapseError(400, "MSISDN is already in use", Codes.THREEPID_IN_USE)
 
+        if not self.hs.config.account_threepid_delegate_msisdn:
+            logger.warn(
+                "No upstream msisdn account_threepid_delegate configured on the server to "
+                "handle this request"
+            )
+            raise SynapseError(
+                400,
+                "Adding phone numbers to user account is not supported by this homeserver",
+            )
+
         ret = yield self.identity_handler.requestMsisdnToken(
-            id_server, country, phone_number, client_secret, send_attempt, next_link
+            self.hs.config.email_account_threepid_delegate_msisdn,
+            country,
+            phone_number,
+            client_secret,
+            send_attempt,
+            next_link,
         )
+
         return 200, ret
 
 
+class AddThreepidSubmitTokenServlet(RestServlet):
+    """Handles 3PID validation token submission for adding a 3PID to a user's account"""
+
+    PATTERNS = client_patterns(
+        "/add_threepid/(?P<medium>[^/]*)/submit_token/*$", releases=(), unstable=True
+    )
+
+    def __init__(self, hs):
+        """
+        Args:
+            hs (synapse.server.HomeServer): server
+        """
+        super(AddThreepidSubmitTokenServlet, self).__init__()
+        self.hs = hs
+        self.auth = hs.get_auth()
+        self.config = hs.config
+        self.clock = hs.get_clock()
+        self.store = hs.get_datastore()
+
+    @defer.inlineCallbacks
+    def on_GET(self, request, medium):
+        # We currently only handle threepid token submissions for email
+        if medium != "email":
```

Comment:
wouldn't it be easier just to match `email` in the URL pattern?

Reply:
Well, we will eventually add support for msisdn, just not yet. We should still handle requests from those that expect it gracefully though.

Reply:
Is responding with `M_UNRECOGNISED` (ie, what would happen if you matched on email in the url pattern) instead of `M_UNKNOWN` (what happens here) insufficently graceful?

Reply:
Mm, well we can respond with `M_UNRECOGNIZED` using the SynapseError as well. I just wanted people to know the medium isn't support vs. they've hit a wrong endpoint.

Reply:
I've switched it to email-only. In the future we should abstract the non-email stuff into it's own method and then call that from this and a dedicated msisdn submit_token endpoint.